### ISSUE TITLE
Bug: sam package overwrites / racy when running in parallel

### ISSUE THREAD

**Issue by Manouchehri** [aws/aws-sam-cli#4676]
https://github.com/aws/aws-sam-cli/blob/ee23953806be066ebfb72dd42e188de4866483ae/samcli/lib/package/utils.py#L211

I'm using `sam package` in parallel to upload and sign my Lambda package into multiple regions. It looks like `utils.py` never accounted for this situation: multiple processes in parallel write to the same file, causing corrupted/truncated ZIP files to be uploaded to S3. (For now, I'm doing `terraform apply -parallelism=1` to workaround the issue, but it's not ideal.)

There's two solutions that I/we could do:

1. Just use `tempfile.mkdtemp` instead. https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp <- my preference
2. Add some code to implement a file lock, and reuse the zip file. (This feels more prone to error.)

**Comment by Manouchehri** [comment]
It's for the same template, but each region has its own signature. e.g. here's an example of signing for both us-east-2 and us-east-1.

```
sam package --region us-east-2 --signing-profiles 'FunctionARM=signer1 FunctionIntel=signer1' --template-file .sambuild/template.yaml --s3-bucket sam-us-east-2 --output-template-file .us-east-2.sam-stack-output.yaml
```

```
sam package --region us-east-1 --signing-profiles 'FunctionARM=signer1 FunctionIntel=signer1' --template-file .sambuild/template.yaml --s3-bucket sam-us-east-1 --output-template-file .us-east-1.sam-stack-output.yaml
```

Does this make sense? =)

**Comment by jfuss** [comment]
Ahh. Multiple regions. That seems reasonable to me. I am not sure if we are doing `gettempdir` for some specific reason. 

Looking at the code, I can't see any that may cause issues. The one thing that comes to mind is that we would want to make sure the same code gets a consistent hash. Otherwise, we will upload copies that already exist, exploding the number of uploaded artifacts. I don't think this will happen since we already computed the hash. 

@sriram-mv Might have deeper knowledge here.

@Manouchehri Are you willing to submit a patch? Moving to `mkdtemp` seems reasonable, assuming there aren't other affects that happen.

**Comment by mndeveci** [comment]
Related: https://github.com/aws/aws-sam-cli/issues/2270

**Comment by mndeveci** [comment]
Released with [v1.82.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.82.0)
